kakao/src/main/kotlin/com/agoda/kakao/intercept/Interceptor.kt

```diff
@@ -0,0 +1,141 @@
+package com.agoda.kakao.intercept
+
+import androidx.test.espresso.DataInteraction
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.ViewAssertion
+import androidx.test.espresso.ViewInteraction
+import androidx.test.espresso.web.assertion.WebAssertion
+import androidx.test.espresso.web.model.Atom
+import androidx.test.espresso.web.sugar.Web
+
+/**
+ * Base class for intercepting the call chain from Kakao to Espresso.
+ *
+ * Interceptors can be provided through [Kakao][com.agoda.kakao.Kakao] runtime,
+ * different [Screens][com.agoda.kakao.screen.Screen] as well as [KViews][com.agoda.kakao.common.views.KBaseView].
+ *
+ * Interceptors are stacked during the runtime for any Kakao-Espresso `check` and `perform` operations.
+ * The stack ordering is following: KView interceptor -> Screen interceptor -> Kakao interceptor.
+ *
+ * Any of the interceptors in the chain can break the chain call, this will not only prevent underlying
+ * interceptors from being invoked, but prevents Kakao from executing the operation. In that case,
+ * responsibility for actually making Espresso call lies on developer.
+ *
+ * For each operation the interceptor invocation cycle will be as follows:
+ * ```
+ * // For check operation
+ * onAll?.invoke()
+ * onCheck?.invoke()
+ *
+ * // For perform operation
+ * onAll?.invoke()
+ * onPerform?.invoke()
+ * ```
+ *
+ * @see com.agoda.kakao.Kakao
+ * @see com.agoda.kakao.screen.Screen
+ * @see com.agoda.kakao.common.views.KBaseView
+ */
+class Interceptor<T, C, P>(
+        val onCheck: Pair<Boolean, (T, C) -> Unit>?,
```

Comment:
Maybe a data class can be used for this instead of Pair? For readability sake. For all onCheck, onPerform and onAll properties.

Reply:
agree

Reply:
Will be pushed soon